Add promoted card to in-App marketplace page

corsonr opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

The search on `WooCommerce > Marketplace` page that shows product from the WooCommerce.com marketplace as cards. During previous re-design phase we added updated product cards in #30410.

This PR adds a new style of "promoted" product cards by:

- adding a new style card
- adding a top border to the card (the color is defined in the JSON feed)
- adding a colored button (the color is defined in the JSON feed)
- adding an automated logic that changes the button text color base on the button background-color

Closes Automattic/woocommerce.com#10885

### Screenshots

![](https://d.pr/i/LmuoA7+)

### How to test the changes in this Pull Request:

- Check out this branch.
- Edit `includes/admin/class-wc-admin-addons.php`.
- On line 66, change `$featured = get_transient( 'wc_addons_featured' );` to bypass the transient check:

```
$featured = false; // TODO - REMOVE THIS
```

- On line 83, change `https://woocommerce.com/wp-json/wccom-extensions/2.0/featured' . $parameter_string` to point to our mock WCCOM featured API data: `https://gist.githubusercontent.com/corsonr/6ac9a36583dce3a2d2d9e97f4ddf7847/raw/60a38e52de705566bdb74b2059d6acc6a086379b/featured.json`.

```php
$raw_featured = wp_safe_remote_get(
	'https://gist.githubusercontent.com/andfinally/be6f6caae34f3f3c1c5de1df0ba30681/raw/featured.json',
	array(
		'headers'    => $headers,
		'user-agent' => 'WooCommerce Addons Page',
	)
);
```

the headers could have to be removed from the `wp_safe_remote_get` call.

- On line 142, change `'https://woocommerce.com/wp-json/wccom-extensions/1.0/search' . $parameters,` to point to another mock JSON endpoint: `https://gist.githubusercontent.com/andfinally/cdea0c85da10bc037dad6aaa8bf28d0e/raw/search.json`.

```php
$raw_extensions = wp_safe_remote_get(
	'https://gist.githubusercontent.com/andfinally/cdea0c85da10bc037dad6aaa8bf28d0e/raw/search.json',
	array( 'headers' => $headers )
);
```

*  View the Featured page of WooCommerce > Marketplace on your WooCommerce site `wp-admin/admin.php?page=wc-addons&section=_featured`.
* In the `Discover our favorites` group you should see the two new styles of card shown in this screenshot:

<img width="750" alt="image" src="https://user-images.githubusercontent.com/1647564/137333981-1ee3a9e4-0069-4920-8b35-a988568382d5.png">

* View a category page like `wp-admin/admin.php?page=wc-addons&section=marketing-extensions` or search results page. You should see the two styles of card shown here at the top:

<img width="750" alt="image" src="https://user-images.githubusercontent.com/1647564/137336982-35aa77a8-ccfa-4519-84f1-87a1411a8623.png">

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Tweak - Add promoted cards styling to marketplace section. #30861

### FOR PR REVIEWER ONLY:

* [x] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

joshbetz committed:
Add mobile data to WCTracker

Adds an endpoint to consolidate the data that is needed every time the
app is opened instead of relying on multiple endpoints.

This is also a good place to include some basic mobile data usage in
WCTracker. We track each platform independently, so we can see which
versions of the apps are used on a given site.

joshbetz committed:
Convert to /telemetry endpoint

Instead of overloading the application startup endpoint with recording
telemetry data, we'll start with a dedicated telemetry endpoint. This
should be easier to reason about and the additional overhead of a
separate endpoint is minimal.

joshbetz committed:
Convert to POST endpoint

We're only writing data now, so it's more natural for this to be a POST
endpoint. We'll send the data in the request body instead of as query
parameters.

joshbetz committed:
Use gmdate instead of date per coding standards

gmdate always returns a GMT date, so is not suspectible to runtime
timezone changes

joshbetz committed:
Fix function name

We were calling this function with a different name

joshbetz committed:
Add argument validation on telemetry endpoint

Properly validate the platform and version parameters in the request
body for telemetry requests.

joshbetz committed:
Add e2e tests for telemetry endpoint

* Verify that missing parameters will result in errors
* Verify that passing correct parameters will result in 200

joshbetz committed:
Additional invalid tests cases

Use it.each to run the same test with various invalid test data

joshbetz committed:
lint

tgglv committed:
Add stars and reviews number

tgglv committed:
Add stars and reviews number

tgglv committed:
Use icon instead of image if possible

tgglv committed:
Use icon instead of image if possible

tgglv committed:
Add developed-by string

tgglv committed:
Add developed-by string

tgglv committed:
Show ratings with the use of stars

tgglv committed:
Show ratings with the use of stars

corsonr committed:
Add promoted card to in-App marketplace page

corsonr committed:
Added full width and text color support

corsonr committed:
Added hover effect on buttons

corsonr committed:
fix PHPCS warnings

corsonr committed:
Fix  remaining PHPCS warning

andfinally commented:
I'm currently having trouble getting the API request to work locally with my [Woo Core development environment site](https://github.com/Automattic/woocommerce-core-development-environment) – I'm getting `cURL error 7: Failed to connect to woocommerce.test port 443: Connection refused` despite having a filter to set the correct HTTP options. I'll try again later!

helgatheviking committed:
Add woocommerce_product_options_shipping_product_data hook to metabox. Closes #30811.

joshbetz committed:
Fix the permission check for telemetry data

Any logged in user should be able to post telemetry data

joshbetz committed:
Fix the docblock copy pasta

corsonr commented:
@andfinally I'm testing from the test customer site if that helps 😉

tgglv committed:
Remove the current addons implementation

tgglv committed:
Address review comments

tgglv committed:
Address review comments

corsonr committed:
Reduce Storefront banner width & track links

corsonr committed:
Fix banner width

tgglv committed:
Refactor product card to use with groups

tgglv committed:
Handle product-like banners

tgglv committed:
Render buttons based banner

tgglv committed:
Output products according to the group's capacity

tgglv committed:
Fix showing rating/reviews number

tgglv committed:
Fix product-related banner's image position

tgglv committed:
Apply product group's capacity

tgglv committed:
Fix icon/image/paragraph of the product card

tgglv committed:
Adjust group's title+description+see more block

tgglv committed:
Use 3-column groups for screens more than 960px

tgglv committed:
Use currency code from WCCOM instead of symbol

tgglv committed:
Revert debug code

madeincosmos commented on includes/admin/class-wc-admin-addons.php:
```diff
@@ -33,7 +33,7 @@ public static function get_featured() {
 
 			$raw_featured = wp_safe_remote_get(
-				'https://woocommerce.com/wp-json/wccom-extensions/1.0/featured',
+				'https://woocommerce.test/wp-json/wccom-extensions/1.0/featured',
```
Let's leave this one out of production :)

corsonr replied:
good catch!

tgglv replied:
It looks like `woocommerce.test` in the URL is still there. Please don't forget to use `woocommerce.com` before the final merge.

madeincosmos commented on assets/css/admin.scss:
```diff
@@ -726,13 +741,34 @@
+						top: 28px;
+						text-align: center;
+						background: #DCDCDE;
+						border-radius: 2px;
```
a minor thing, but it would be good to arrange these properties alphabetically like the linter suggests

madeincosmos reviewed (commented):
Thank you for the PR @corsonr! Everything looks good to me except for the test site URL that accidentally made its way to the commit, we'll need to replace it with the original feed before merging. Other than that, the cards are looking great on all screen sizes 🌟

lanej0 committed:
Quote WORDPRESS_TITLE value

tgglv commented on includes/admin/class-wc-admin-addons.php:
```diff
@@ -305,6 +305,40 @@ public static function output_column_block( $block ) {
+		<ul class="products">
+			<?php
+			foreach ( $block->items as $card ) {
+				if ( true === $card->full_width ) {
```
It seems that checking agains some kind of `isset` would be vital here. Currently with data from the PR's testing instructions I have to following notices:
![image](https://user-images.githubusercontent.com/15897544/136790129-93ffe9f3-52bc-4763-90c1-5638bd7b18bf.png)

Perhaps something like the following will fit us best (sorry, I'm syntactic sugar addict :yum:):
```php
$full_width = $card->full_width ?? false;
if ( true === $full_width ) {
    ...
```

tgglv reviewed (commented):
Thanks, @corsonr!

It works like a charm! 🍀 

I've made it as a comment rather than approval to ask you to apply Maria's and mine suggestions first.

corsonr committed:
Fix CSS lints & Update remote URL

joshbetz committed:
Change endpoint to wc-telemetry/tracker

Since this endpoint isn't a publicly documented endpoint, we'll keep it
out of the wc/v3 namespace. Instead we're introducing a new wc-telemetry
namespace. The name of the endpoint then is updating to tracker.

The wc-telemetry endpoint is unversioned for consistency with
wc-analytics and because it's not publicly documented.

corsonr committed:
Place Storefront URL into a single variable

joshbetz committed:
Add docblock for get_telemetry_controllers

madeincosmos reviewed (dismissed):
Thank you for the fixes, everything looks good to me now on all screen sizes 🎉

andfinally committed:
Changed SCSS comments.

corsonr committed:
Fix banner width for small devices

andfinally committed:
Added UTM params to vendor URL.

andfinally committed:
Added support for dynamic price suffix in `$addon` data.

tgglv committed:
Test: clean up before running tests

tgglv committed:
Try remove vendor folder from cache

tgglv committed:
Try remove vendor folder from caching

tgglv committed:
Try remove using of cache

tgglv committed:
Try add phpunit into deps

tgglv committed:
Fix linter bugs

joshbetz committed:
Update tests to use new endpoint

tgglv committed:
Move get_star_class out of template file

tgglv committed:
Revert composer.json

andfinally committed:
Update includes/admin/class-wc-admin-addons.php

Returning country in request to featured 2.0 endpoint.

andfinally committed:
Update includes/admin/class-wc-admin-addons.php

Adding `$parameter_string` to end of URL for featured endpoint.

andfinally committed:
Updated content-hash to try and fix phpunit failure. https://github.com/woocommerce/woocommerce/pull/30840\#issuecomment-942422984.

andfinally committed:
Merge branch 'trunk' into update/marketplace-featured-page

# Conflicts:
#	includes/admin/class-wc-admin-addons.php
#	includes/admin/views/html-admin-page-addons.php

andfinally committed:
Reverting changes to CircleCi config in 618eca7c6ad76a93bb61972c0bdb55eddff03317 and cc0b78f8ecf878ebc74a20a49c5e29045432050e. These were attempts to unstick PHPUnit tests.

andfinally committed:
Added condition to prevent showing `Sorry, could not find anything` message on featured page. This should only be shown when a search term has been submitted.

andfinally committed:
Restored `output_search_promotion_block` and some of the CSS for `.addons-wcs-banner-block`.

andfinally committed:
Restored more styles for `.addons-wcs-banner-block` used for WCPay and WCShip banners.
Restored `install_woocommerce_services_addon` and `install_woocommerce_payments_addon` methods, needed to install plugins from WCPay and WCShip banners.

lanej0 committed:
Update changelog

corsonr committed:
Update content-hash in composer.lock

corsonr committed:
Merge branch 'trunk' into update/wccom-11343

Konamiman committed:
Merge pull request #30882 from woocommerce/update/wccom-11343

Reduce Storefront banner width & track links

tgglv committed:
Apply PHPCS suggestions

andfinally committed:
Restored and annotated deprecated functions for backwards compatibility.

andfinally committed:
Addressing feedback. Using cheaper escaping methods.

andfinally committed:
Correcting escaping of `$mapped->price`, using `wp_kses`.

andfinally committed:
Merge branch 'trunk' into update/marketplace-featured-page

corsonr committed:
Add featured & Promoted Cards to Search Results

This commits addresses the display of the promoted product cards and the featured products cards in the search results  on the marketplace page.

It requires `plugins/wccom-plugins/extensions-api/includes/search/class-database-search.php` to include 4 new variables in `$results[]`:

- label
- primary_color
- text_color
- button

tgglv committed:
Merge remote-tracking branch 'origin/trunk' into add/reviews-into-marketplace-product-card

corsonr committed:
Fix indentation

andfinally committed:
Added bottom margin of paragraph in `.addons-buttons-banner-details`, to improve rendering of promotional banner on featured page.

andfinally committed:
Merge branch 'update/marketplace-featured-page' of https://github.com/woocommerce/woocommerce into update/marketplace-featured-page

tgglv committed:
Merge remote-tracking branch 'origin/update/marketplace-featured-page' into add/wccom-10885

tgglv committed:
Apply linter related changes

tgglv committed:
Revert promoted card output

corsonr committed:
Ensure $addon->label is set in if statement

corsonr committed:
Style featured & promoted product cards in search results

corsonr committed:
Fix PHPCS lints

vedanshujain committed:
Merge pull request #30876 from helgatheviking/issues/30811-shipping-metabox

Add woocommerce_product_options_shipping_product_data hook to metabox

andfinally committed:
Added support for promoted items in featured endpoint.

tgglv committed:
Fix PHPCS issues

Konamiman committed:
Merge pull request #30840 from woocommerce/add/reviews-into-marketplace-product-card

Add ratings, reviews and icons into Marketplace's Product Cards

andfinally commented:
For future reference, this is the structure of a "featured" and a "promoted" product that will be returned from the WCCOM API.

```json
{
    "title": "Stripe",
    "image": "https://woocommerce.com/wp-content/uploads/2020/06/Logo-Stripe-Dark-1.png",
    "excerpt": "Accept all major debit and credit cards as well as local payment methods with Stripe.",
    "link": "https://woocommerce.com/products/stripe/?utm_source=extensionsscreen&utm_medium=product&utm_campaign=wcaddons",
    "demo_url": "",
    "price": "&#36;0.00",
    "hash": "50bb7a985c691bb943a9da4d2c8b5efd",
    "slug": "woocommerce-gateway-stripe",
    "id": 18627,
    "rating": null,
    "reviews_count": null,
    "vendor_name": "WooCommerce",
    "vendor_url": "https://woocommerce.com/vendor/woocommerce/",
    "icon": "https://woocommerce.com/wp-content/uploads/woocommerce_uploads/2011/12/stripe-app-icon-7m1xi7.png",
    "label": "featured"
},
{
    "title": "WooCommerce Payments",
    "image": "https://woocommerce.com/wp-content/uploads/2020/06/Logo-Woo-Pay-Dark.png",
    "excerpt": "Securely accept payments, track cash flow, and manage recurring revenue from your dashboard \u2014 all without setup costs or monthly fees.",
    "link": "https://woocommerce.com/products/woocommerce-payments/?utm_source=extensionsscreen&utm_medium=product&utm_campaign=wcaddons",
    "demo_url": "",
    "price": "&#36;0.00",
    "hash": "8c6319ca-8f41-4e69-be63-6b15ee37773b",
    "slug": "woocommerce-payments",
    "id": 5278104,
    "rating": null,
    "reviews_count": null,
    "vendor_name": "WooCommerce",
    "vendor_url": "https://woocommerce.com/vendor/woocommerce/",
    "icon": "https://woocommerce.com/wp-content/uploads/woocommerce_uploads/2020/02/wcpay-icon-fv4tef.png",
    "primary_color": "#FFE320",
    "text_color": "#2c3338",
    "button": "View details",
    "label": "promoted"
},
```

andfinally committed:
Added link to product name in promoted card, and adjusted size of product icon to 80px.

andfinally reviewed (dismissed):
Looking great, thank you! 👌

Apologies to other reviewers for the whitespace changes.

andfinally committed:
Added border radius to icon on promoted card.

tgglv reviewed (dismissed):
Thanks, @corsonr!

Changes look great and work like a charm 🍀 for the featured page, search results (for both: search by a term and category).

The only thing according to test instructions. I can't test with the GH gist (don't know why). Other than that I've used local WCCOM.

andfinally commented:
Thanks Timur! In case needed, I've pasted the contents of the two gists into these a8c pastebins.

featured.json: 2a72a-pb/
search.json: 2a72b-pb/

jonathansadowski committed:
Merge pull request #30415 from woocommerce/add/wctracker-mobile-data

Add mobile data to WCTracker

rrennick committed:
Merge pull request #30913 from lanej0/fix/quote-string-for-docker-compatibility

Quote WORDPRESS_TITLE value

tgglv committed:
Merge remote-tracking branch 'origin/trunk' into update/marketplace-featured-page

tgglv committed:
Fix after merge with trunk

tgglv committed:
Revert get_featured method

Konamiman committed:
Merge pull request #30900 from woocommerce/update/marketplace-featured-page

Revamp Featured page in WooCommerce Marketplace

tgglv committed:
Merge remote-tracking branch 'origin/trunk' into add/wccom-10885

tgglv reviewed (dismissed):
Thanks, @corsonr!

The code looks great and works like a charm! 🍀 

Tested it once again after resolving conflicts with the `trunk`. Everything works as expected (for featured/promoted/regular product cards for both featured page and search).

andfinally committed:
Deleted unneeded method `output_promoted_card`.

andfinally committed:
Added full stops on inline comments.

corsonr commented:
@Konamiman 

>wp-admin/admin.php?page=wc-addons&section=_featured shows one style of cards, not two

Instructions have been updated, `render_featured()` needs to be updated, since `get_featured()` is deprecated.

I'm getting correct results:

![](https://d.pr/i/ZdBlSS+)

![](https://d.pr/i/CijhUh+)

andfinally committed:
Merge branch 'release/5.9' into add/wccom-10885

andfinally committed:
Updated hash in `composer.lock`.
Fixed `get_star_class` method that was messed up in resolving conflicts in GitHub.

andfinally committed:
Addressing PHPCS warning. Changed `else if` to `elseif` in `get_star_class` method.

andfinally committed:
Tidied up `@deprecated` comments and removed unnecessary whitespace changes.

andfinally committed:
Removing differences against `release/5.9` that are not connected to the changes in this branch.

andfinally committed:
Fixing PHPCS warning. Replaced `else if` with `elseif` in `WC_Admin_Addons::get_star_class`.

andfinally committed:
Adding UTM parameters to vendor link in in-app marketplace product card.

andfinally committed:
Restored `WC_Admin_Addons::output_search_promotion_block` call in In-App Marketplace template page, which was accidentally omitted when fixing conflicts.

corsonr committed:
Add missing whitespace

Konamiman commented:
Additional pull request to incorporate these changes in WooCommerce 6.0: https://github.com/woocommerce/woocommerce/pull/31090
